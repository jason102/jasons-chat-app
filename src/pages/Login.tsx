import { Link, Button } from '@mui/material';
import React from 'react';
import { useNavigate } from 'react-router-dom';
import {
  Form,
  Section,
  TextContainer,
  TextInput,
  SubmitButtonContainer,
  ErrorMessage,
  OtherAuthPageLink,
} from 'components/AuthPages.styles';
import { LoginFormData, useLogin } from 'hooks/authHooks';

const Login: React.FC = () => {
  const { formData, setFormData, handleSubmit } = useLogin();
  const navigate = useNavigate();

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    setFormData((prevFormData: LoginFormData) => ({
      ...prevFormData,
      [e.target.name]: e.target.value,
    }));
  };

  const onGoToRegister = () => {
    navigate('/register');
  };

  const { email, password, error, loading } = formData;

  return (
    <Section>
      <h3>Welcome! Please login:</h3>
      <Form onSubmit={handleSubmit}>
        <TextContainer>
          <TextInput
            label='Email'
            variant='outlined'
            name='email'
            value={email}
            disabled={loading}
            onChange={handleChange}
            autoComplete='email'
          />
        </TextContainer>
        <TextContainer>
          <TextInput
            label='Password'
            variant='outlined'
            name='password'
            type='password'
            value={password}
            onChange={handleChange}
            disabled={loading}
            autoComplete='current-password'
          />
        </TextContainer>
        {error && <ErrorMessage>{error}</ErrorMessage>}
        <SubmitButtonContainer>
          <Button size='large' type='submit' disabled={loading}>
            {loading ? 'Please wait...' : 'Login'}
          </Button>
        </SubmitButtonContainer>
      </Form>
      <OtherAuthPageLink>
        {`Don't have an account? Please `}
        <Link onClick={onGoToRegister}>register</Link>
      </OtherAuthPageLink>
    </Section>
  );
};

export default Login;
