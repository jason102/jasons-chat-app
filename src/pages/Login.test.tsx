import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Login from './Login';

jest.mock('firebase/app');
jest.mock('firebase/auth');
jest.mock('firebase/firestore');

const mockedUseNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...(jest.requireActual('react-router-dom') as any),
  useNavigate: () => mockedUseNavigate,
}));

describe('Login page', () => {
  it('renders', () => {
    const { asFragment } = render(<Login />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should show text in an input field when the user enters text', () => {
    render(<Login />);

    const emailInput = screen.getByRole('textbox', {
      name: /email/i,
    });
    fireEvent.change(emailInput, { target: { value: 'jason@email.com' } });

    expect(emailInput).toHaveValue('jason@email.com');
  });

  it('should show an error if some fields are still blank when submitting the form', () => {
    render(<Login />);

    const loginButton = screen.getByRole('button', {
      name: /login/i,
    });

    userEvent.click(loginButton);

    expect(
      screen.getByText(/all fields must be filled in/i)
    ).toBeInTheDocument();
  });

  it('should navigate to the chat page when the form is successfully submitted', async () => {
    render(<Login />);

    const emailInput = screen.getByRole('textbox', {
      name: /email/i,
    });
    const passwordInput = screen.getByLabelText(/password/i);
    fireEvent.change(emailInput, { target: { value: 'jason@email.com' } });
    fireEvent.change(passwordInput, { target: { value: 'abc123' } });

    const loginButton = screen.getByRole('button', {
      name: /login/i,
    });
    userEvent.click(loginButton);

    await waitFor(() => {
      expect(mockedUseNavigate).toHaveBeenCalledTimes(1);
      expect(mockedUseNavigate).toBeCalledWith('/');
    });
  });

  it('should navigate to the registration page when clicking on the bottom register link', () => {
    render(<Login />);

    const registerLink = screen.getByText(/register/i);

    userEvent.click(registerLink);

    expect(mockedUseNavigate).toHaveBeenCalledTimes(1);
    expect(mockedUseNavigate).toBeCalledWith('/register');
  });
});
