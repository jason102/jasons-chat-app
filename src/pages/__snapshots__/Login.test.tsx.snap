// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login page renders 1`] = `
<DocumentFragment>
  <section
    class="sc-bcXHqe dBnoDu"
  >
    <h3>
      Welcome! Please login:
    </h3>
    <form
      class="sc-gswNZR eXLFlH"
    >
      <div
        class="sc-dkrFOg leTtXt"
      >
        <div
          class="MuiFormControl-root MuiTextField-root sc-hLBbgP cHttSm css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for=":r0:"
            id=":r0:-label"
          >
            Email
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              autocomplete="email"
              class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id=":r0:"
              name="email"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-yjsfm1"
              >
                <span>
                  Email
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
      <div
        class="sc-dkrFOg leTtXt"
      >
        <div
          class="MuiFormControl-root MuiTextField-root sc-hLBbgP cHttSm css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for=":r1:"
            id=":r1:-label"
          >
            Password
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              autocomplete="current-password"
              class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id=":r1:"
              name="password"
              type="password"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-yjsfm1"
              >
                <span>
                  Password
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
      <div
        class="sc-eDvSVe jONXSr"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeLarge MuiButton-textSizeLarge MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeLarge MuiButton-textSizeLarge css-kpm0tx-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="submit"
        >
          Login
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </form>
    <div
      class="sc-gKPRtg fnFZvH"
    >
      Don't have an account? Please 
      <a
        class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1ps4owl-MuiTypography-root-MuiLink-root"
      >
        register
      </a>
    </div>
  </section>
</DocumentFragment>
`;
